CONTRIBUTE.md

```diff
@@ -0,0 +1,156 @@
+# Architecture
+
+Galaxykit has a fairly simple structure:
+
+```
+(galaxykit) henderson@mars => tree .
+.
+├── CHANGES.md
+├── CONTRIBUTE.md
+├── galaxykit
+│   ├── client.py
+│   ├── collections.py
+│   ├── command.py
+│   ├── container_images.py
+│   ├── containers.py
+│   ├── containerutils.py
+│   ├── groups.py
+│   ├── namespaces.py
+│   ├── registries.py
+│   └── users.py
+├── LICENSE.md
+├── README.md
+├── setup.cfg
+└── setup.py
+```
+
+Let's work through these files individually:
+
+### CHANGES.md
+A changelog, tracking all the updates to the code.
+
+### CONTRIBUTE.md
+
+This file, with advice on how to effectively contribute to galaxykit.
+
+### galaxykit:
+
+The source directory, containing all the code that goes into galaxykit.
+
+#### client.py
+
+This file contains the GalaxyClient object, which is essentially a single authenticated context for making requests against an existing galaxy_ng instance. This file and `command.py` contain the primary two interfaces to interacting with galaxykit.
+
+#### collections.py
+
+Functions for managing collections.
+
+#### command.py
+
+The command-line interface logic - this is what gets run when you invoke galaxykit directly on the command-line. Currently in a bit of a hacky state, as it's grown faster than we can refactor it. An extensive refactor is in progress - check out [this PR](https://github.com/ansible/galaxykit/pull/30) if you want to review the changes.
+
+#### container_images.py
+
+For interacting with individual container images.
+
+#### containers.py
+
+Functions for dealing with containers as as discrete units (deleting, setting metadata.)
+
+#### containerutils.py
+
+This code abstracts over common `podman` or `docker` commands for uploading or downloading container images from a galaxy_ng instance.
+
+#### groups.py
+
+Adding/deleting/modifying the permissions of groups.
+
+#### namespaces.py
+
+Creating and deleting namespaces.
+
+#### registries.py
+
+Adding new remote container registries.
+
+#### users.py
+
+Adding/deleting users.
+
+### LICENSE.md
+
+This is the GPLv2, which is the same licence galaxy_ng is distributed under.
+
+### README.md
+
+Landing page.
+
+### setup.cfg
+
+Currently just contains a few linter settings.
+
+### setup.py
+
+All the metadata and things used when releasing this utility on PyPi.
+
+
+# Development Practices
+
+Please:
+
+1. Ensure that your PR has a changelog addition for any features or bugs fixed that are contained in your PR.
```

Comment:
Looks like previously this only happened on release.

Should we requiring every PR to prepend an entry to this file?  Because this would create a merge conflict whenever a PR is merged, as they would all be touching line 1.

I wonder if we should just require changelog-worthy PR titles,
and populate the changelog from `git log v0.5.2... --format=%s | tac` on release.

Reply:
It had happened only on release because I was bad about remembering to add things in progressively :sweat_smile: 

I think that the approach you mention is perhaps the right way to do it - is that similar to how it's done in the galaxy_ng and ansible-hub-ui repositories? I'm actually coming around to the idea that we should probably not differ too much from how it's done there, and that way there won't be too much to remember when developing in different repositories.

Reply:
I'm going to remove the PR requirement from this guide, and am going to add a ticket to Jira to get the changelog addition worked out.

Reply:
>  is that similar to how it's done in the galaxy_ng and ansible-hub-ui repositories?

Well, the principle is the same, but in the UI we have the `pr_checks` script which requires a `CHANGES/123.abc` file describing the change .. and then those get concatenated into `CHANGES.md` by `towncrier`.

(I still populate that file by doing `git log -1 --format=%s > CHANGES/...`, but we do have that extra file step...)